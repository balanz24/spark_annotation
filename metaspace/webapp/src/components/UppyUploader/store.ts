// based on https://uppy.io/docs/stores/#Implementing-Stores

import { ref } from '@vue/composition-api'
import { State } from '@uppy/core'

type Listener = (prevState: State, nextState: State, patch: object) => void

export default () => {
  const state = ref<State>({})
  const listeners = new Set<Listener>()

  return {
    getState: () => state.value,
    setState: (patch: object) => {
      const prevState = state.value
      const nextState = { ...prevState, ...patch }

      state.value = nextState

      listeners.forEach(listener => {
        listener(prevState, nextState, patch)
      })
    },
    subscribe: (listener: Listener) => {
      listeners.add(listener)
      return () => listeners.delete(listener)
    },
  }
}
