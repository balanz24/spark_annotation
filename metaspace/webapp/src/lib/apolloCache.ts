import { defaultDataIdFromObject, InMemoryCache } from 'apollo-client-preset'

const nonNormalizableTypes: any[] = ['User', 'DatasetUser', 'DatasetGroup', 'DatasetProject']
export const makeApolloCache = () => new InMemoryCache({
  cacheRedirects: {
    Query: {
      // Allow get-by-id queries to use cached data that originated from other kinds of queries
      dataset: (_, args, { getCacheKey }) => getCacheKey({ __typename: 'Dataset', id: args.id }),
      annotation: (_, args, { getCacheKey }) => getCacheKey({ __typename: 'Annotation', id: args.id }),
      group: (_, args, { getCacheKey }) => getCacheKey({ __typename: 'Group', id: args.groupId }),
      project: (_, args, { getCacheKey }) => getCacheKey({ __typename: 'Project', id: args.projectId }),
    },
  },
  dataIdFromObject(object: any) {
    // WORKAROUND: Because of Apollo's aggressive caching, often the current User will be overwritten with results
    // from other queries. The server side often strips fields based on how they're accessed (the "ScopeRole" logic),
    // which means these query paths will often return different data with the same IDs:
    // currentUser -> primaryGroup (always present)
    // dataset -> submitter -> primaryGroup (null unless admin)
    // To protect against this, don't allow Users (and possibly other types in the future) to have a dataId,
    // so that InMemoryCache cannot share data between different queries.
    if (nonNormalizableTypes.includes(object.__typename)) {
      return null
    } else {
      return defaultDataIdFromObject(object)
    }
  },
})
