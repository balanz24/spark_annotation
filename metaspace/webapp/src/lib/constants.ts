export const MAX_MOL_DBS = 4
export const MAX_MOL_DBS_EXT = 6
export const MAX_NEUTRAL_LOSSES = 3
export const MAX_CHEM_MODS = 3
export const DEFAULT_SCALE_TYPE = 'linear'
