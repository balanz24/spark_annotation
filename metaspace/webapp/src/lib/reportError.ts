import { ElNotification } from 'element-ui/types/notification'
import * as Sentry from '@sentry/browser'
import { every } from 'lodash-es'
import { Primitive } from 'ts-essentials'

let $notify: ElNotification

export function setErrorNotifier(_$notify: ElNotification) {
  $notify = _$notify
}

function isHandled(err: any) {
  try {
    return !!((err?.graphQLErrors && every(err.graphQLErrors, e => e && e.isHandled)) || err?.isHandled)
  } catch {
    return false
  }
}

const DEFAULT_MESSAGE = 'Oops! Something went wrong. Please refresh the page and try again.'

export default function reportError(
  err: Error,
  message: string | null = DEFAULT_MESSAGE,
  extraData?: Record<string, Primitive>,
) {
  try {
    if (!isHandled(err)) {
      Sentry.captureException(err, { contexts: { 'Extra Data': extraData } })
      console.error(err, extraData)
      if ($notify != null && message) {
        $notify.error(message)
      }
    }
  } catch (ex) {
    console.error(ex)
    /* Avoid breaking down-stream error handling  */
  }
}
