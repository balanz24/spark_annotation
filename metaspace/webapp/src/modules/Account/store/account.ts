import { Module } from 'vuex'
import { DialogType } from '../dialogs'
import router from '../../../router'
import { RawLocation } from 'vue-router/types/router'

export interface AccountState {
  dialog: DialogType | null;
  dialogCloseRedirect: RawLocation | null;
  loginSuccessRedirect: RawLocation | null;
}

interface ShowDialogOptions {
  dialog: DialogType;
  dialogCloseRedirect?: RawLocation;
  loginSuccessRedirect?: RawLocation;
}

interface HideDialogOptions {
  dialog: DialogType;
  isLoginSuccess?: boolean;
}

const account: Module<AccountState, any> = {
  namespaced: true,
  state: {
    dialog: null,
    dialogCloseRedirect: null,
    loginSuccessRedirect: null,
  },
  mutations: {
    showDialog(state, payload: DialogType | ShowDialogOptions) {
      const options: ShowDialogOptions = typeof payload === 'string' ? { dialog: payload } : payload

      state.dialog = options.dialog
      if (options.dialogCloseRedirect != null) {
        state.dialogCloseRedirect = options.dialogCloseRedirect
      }
      if (options.loginSuccessRedirect != null) {
        state.loginSuccessRedirect = options.loginSuccessRedirect
      }
    },
    hideDialog(state, payload: DialogType | HideDialogOptions) {
      const options: HideDialogOptions = typeof payload === 'string' ? { dialog: payload } : payload

      if (state.dialog === options.dialog) {
        state.dialog = null
        if (options.isLoginSuccess && state.loginSuccessRedirect != null) {
          router.push(state.loginSuccessRedirect)
        } else if (state.dialogCloseRedirect != null) {
          router.push(state.dialogCloseRedirect)
        }

        state.dialogCloseRedirect = null
        state.loginSuccessRedirect = null
      }
    },
  },
}

export default account
