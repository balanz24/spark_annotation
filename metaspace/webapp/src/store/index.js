import Vue from 'vue';
import Vuex from 'vuex';
Vue.use(Vuex);

import getters from './getters.js';
import mutations from './mutations.js';
import actions from './actions.js';
import {accountModule} from '../modules/Account';

const store = new Vuex.Store({
  state: {
    // names of currently shown filters
    orderedActiveFilters: [],

    filterLists: null,
    filterListsLoading: false,

    // currently selected annotation
    annotation: undefined,

    // currently selected normalization matrix
    normalization: undefined,

    // roi settings
    roiInfo: {visible: false},

    // is annotation table loading?
    tableIsLoading: true,

    lastUsedFilters: {},

    currentTour: null,

    currentUser: {},

    // ion image global viewer settings
    channels: [],
    mode: 'SINGLE'
  },

  getters,
  mutations,
  actions,
  modules: {
    account: accountModule
  }
})

export default store;
