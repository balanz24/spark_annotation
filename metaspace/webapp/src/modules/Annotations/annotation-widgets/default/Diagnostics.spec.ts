import { mount } from '@vue/test-utils'
import router from '../../../../router'
import { initMockGraphqlClient, apolloProvider } from '../../../../../tests/utils/mockGraphqlClient'
import Diagnostics from './Diagnostics.vue'
import store from '../../../../store/index'
import Vue from 'vue'
import Vuex from 'vuex'
import { sync } from 'vuex-router-sync'
import * as reportErrorModule from '../../../../lib/reportError'

Vue.use(Vuex)
sync(store, router)

describe('Diagnostics', () => {
  const peakChartData = JSON.stringify({
    mz_grid: { min_mz: 99.9, max_mz: 101.1 },
    theor: {
      centroid_mzs: [100, 101],
      mzs: [99.9, 100, 100.1, 100.9, 101, 101.1],
      ints: [0, 1, 0, 0, 0.5, 0],
    },
    ppm: 3,
    sampleData: {
      mzs: [100, 101],
      ints: [1, 0.6],
    },
  })

  const metrics = {
    chaos: 0.9976041666666666,
    spatial: 0.005921380836044965,
    spectral: 0.7906196869421936,
  }
  const metricsJson = JSON.stringify(metrics)
  const baseAnnotation = {
    id: '2019-10-15_17h02m31s_7277991',
    ion: 'C19H18N2O7S2-H+',
    ionFormula: 'C19H19N2O7S2',
    mz: 449.0482267014823,
    fdrLevel: 0.1,
    msmScore: 0.657486,
    rhoSpatial: 0.686362,
    rhoSpectral: 0.959122,
    rhoChaos: 0.998756,
    peakChartData,
    possibleCompounds: [
      { name: 'C.I. Food Red 6', information: [{ databaseId: 'HMDB0032738' }] },
    ],
    dataset: {
      id: '2019-02-12_15h55m06s',
      name: 'Untreated_3_434',
      scoringModel: null,
      configJson: JSON.stringify({
        fdr: { scoring_model: null },
      }),
    },
    databaseDetails: {
      id: 24,
    },
    offSample: false,
    offSampleProb: 0.03,
    isobars: [
      {
        ion: 'C8H10S2-H-',
        ionFormula: 'C8H9S2',
        peakNs: [[1, 1], [3, 3]],
        shouldWarn: true,
      },
    ],
    isotopeImages: [
      {
        mz: 100,
        url: '/fs/iso_images/img1',
        minIntensity: 0,
        maxIntensity: 100,
        totalIntensity: 5000,
      },
      {
        mz: 101,
        url: '/fs/iso_images/img2',
        minIntensity: 100,
        maxIntensity: 200,
        totalIntensity: 10000,
      },
    ],
    metricsJson,
  }
  const isobarAnnotation = {
    ...baseAnnotation,
    id: '15_17h02m31s_7277992',
    ion: 'C8H10S2-H-',
    ionFormula: 'C8H9S2',
    offSample: true,
    offSampleProb: 0.87654,
  }

  const props = {
    annotation: baseAnnotation,
    colormap: 'Viridis',
    imageLoaderSettings: {
      annotImageOpacity: 1,
      opacityMode: 'constant',
      imagePosition: { zoom: 1, xOffset: 0, yOffset: 0 },
      opticalSrc: null,
      opticalTransform: null,
      pixelAspectRatio: 1,
    },
  }

  const stubs = ['ImageLoader']

  const W = 200
  const H = 300

  beforeEach(() => {
    // Set HTMLElements to have non-zero dimensions
    // @ts-ignore
    jest.spyOn(HTMLElement.prototype, 'getBoundingClientRect').mockImplementation(() =>
      ({ left: 200, right: 200 + W, top: 100, bottom: 100 + H, width: W, height: H }))
    jest.spyOn(HTMLElement.prototype, 'clientWidth', 'get').mockImplementation(() => W)
    jest.spyOn(HTMLElement.prototype, 'clientHeight', 'get').mockImplementation(() => H)
  })

  it('should match snapshot (no isobars)', async() => {
    initMockGraphqlClient({
      Query: () => ({
        allAnnotations: () => ([]),
        annotation: ({ id }: {id: string}) => ({ ...baseAnnotation, id, peakChartData }),
      }),
    })
    const propsData = {
      ...props,
      annotation: {
        ...baseAnnotation,
        isobars: [],
      },
    }
    const wrapper = mount(Diagnostics, { store, router, apolloProvider, stubs, propsData })
    await Vue.nextTick()

    expect(wrapper).toMatchSnapshot()
  })

  it('should match snapshot (with isobar)', async() => {
    initMockGraphqlClient({
      Query: () => ({
        allAnnotations: () => ([isobarAnnotation]),
        annotation: (_: any, { id }: {id: string}) => ({
          ...(id === isobarAnnotation.id ? isobarAnnotation : baseAnnotation),
          id,
          peakChartData,
        }),
      }),
    })
    const reportErrorFunc = jest.spyOn(reportErrorModule, 'default')

    const wrapper = mount(Diagnostics, { store, router, apolloProvider, stubs, propsData: props })
    await Vue.nextTick()
    wrapper.setData({ comparisonIonFormula: isobarAnnotation.ionFormula })
    await Vue.nextTick()

    expect(reportErrorFunc).not.toBeCalled() // "Inconsistent annotations" warning should fail the test
    expect(wrapper).toMatchSnapshot()
  })
})
