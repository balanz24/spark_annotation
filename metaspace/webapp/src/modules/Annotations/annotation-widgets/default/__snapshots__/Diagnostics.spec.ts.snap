// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diagnostics should match snapshot (no isobars) 1`] = `
<div mock-v-loading="0">
  <!---->
  <!---->
  <!---->
  <div class="">
    <div class="el-row" id="scores-table">
      <div class="msm-score-calc" mock-v-loading="1">MSM score = <span>0.657</span> = 0.006 (ρ<sub>spatial</sub>) × 0.791 (ρ<sub>spectral</sub>) × 0.998 (ρ<sub>chaos</sub>)</div>
      <div>
        <mock-el-popover trigger="hover" open-delay="100">
          <div slot-key="default">Image analysis gave an off-sample probability of less than 10%.</div>
          <div slot-key="reference"><span class="on-sample-tag">On-sample</span></div>
        </mock-el-popover>
      </div>
    </div>
    <div class="isotope-images-container tour-isotope-images el-row">
      <div class="el-col el-col-24 el-col-xs-24 el-col-sm-12 el-col-md-12 el-col-lg-6">
        <div class="small-peak-image">
          100.0000
          <!----> <br>
          <imageloader-stub src="/fs/iso_images/img1" colormap="Viridis" image-fit-params="[object Object]" min-intensity="0" max-intensity="100" show-pixel-intensity="" annotimageopacity="1" opacitymode="constant" imageposition="[object Object]" pixelaspectratio="1" style="overflow: hidden;"></imageloader-stub>
        </div>
      </div>
      <div class="el-col el-col-24 el-col-xs-24 el-col-sm-12 el-col-md-12 el-col-lg-6">
        <div class="small-peak-image">
          101.0000
          <!----> <br>
          <imageloader-stub src="/fs/iso_images/img2" colormap="Viridis" image-fit-params="[object Object]" min-intensity="100" max-intensity="200" show-pixel-intensity="" annotimageopacity="1" opacitymode="constant" imageposition="[object Object]" pixelaspectratio="1" style="overflow: hidden;"></imageloader-stub>
        </div>
      </div>
    </div>
  </div>
  <!---->
  <!---->
  <div>
    <div class="el-row" id="isotope-plot-container">
      <div class="peak-chart">
        <div style="height: 300px;"><svg width="200" height="300">
            <defs>
              <pattern id="pattern-stripe" width="6" height="6" patternUnits="userSpaceOnUse" patternTransform="rotate(45)">
                <rect width="3" height="6" transform="translate(0,0)" fill="white"></rect>
              </pattern>
              <mask id="mask-stripe">
                <rect x="0" y="0" width="100%" height="100%" fill="url(#pattern-stripe)"></rect>
              </mask>
            </defs>
            <g transform="translate(40, 10)">
              <g transform="translate(0, 240)" fill="none" font-size="10" font-family="sans-serif" text-anchor="middle">
                <path class="domain" stroke="currentColor" d="M0.5,6V0.5H120.5V6"></path>
                <g class="tick" opacity="1" transform="translate(10.499999999999526,0)">
                  <line stroke="currentColor" y2="6"></line><text fill="currentColor" y="9" dy="0.71em">100.0</text>
                </g>
                <g class="tick" opacity="1" transform="translate(30.5,0)">
                  <line stroke="currentColor" y2="6"></line><text fill="currentColor" y="9" dy="0.71em">100.2</text>
                </g>
                <g class="tick" opacity="1" transform="translate(50.500000000000476,0)">
                  <line stroke="currentColor" y2="6"></line><text fill="currentColor" y="9" dy="0.71em">100.4</text>
                </g>
                <g class="tick" opacity="1" transform="translate(70.49999999999953,0)">
                  <line stroke="currentColor" y2="6"></line><text fill="currentColor" y="9" dy="0.71em">100.6</text>
                </g>
                <g class="tick" opacity="1" transform="translate(90.5,0)">
                  <line stroke="currentColor" y2="6"></line><text fill="currentColor" y="9" dy="0.71em">100.8</text>
                </g>
                <g class="tick" opacity="1" transform="translate(110.50000000000047,0)">
                  <line stroke="currentColor" y2="6"></line><text fill="currentColor" y="9" dy="0.71em">101.0</text>
                </g>
              </g>
              <g fill="none" font-size="10" font-family="sans-serif" text-anchor="end">
                <path class="domain" stroke="currentColor" d="M-6,240.5H0.5V0.5H-6"></path>
                <g class="tick" opacity="1" transform="translate(0,240.5)">
                  <line stroke="currentColor" x2="-6"></line><text fill="currentColor" x="-11" dy="0.32em">0</text>
                </g>
                <g class="tick" opacity="1" transform="translate(0,192.5)">
                  <line stroke="currentColor" x2="-6"></line><text fill="currentColor" x="-11" dy="0.32em">20</text>
                </g>
                <g class="tick" opacity="1" transform="translate(0,144.5)">
                  <line stroke="currentColor" x2="-6"></line><text fill="currentColor" x="-11" dy="0.32em">40</text>
                </g>
                <g class="tick" opacity="1" transform="translate(0,96.5)">
                  <line stroke="currentColor" x2="-6"></line><text fill="currentColor" x="-11" dy="0.32em">60</text>
                </g>
                <g class="tick" opacity="1" transform="translate(0,48.5)">
                  <line stroke="currentColor" x2="-6"></line><text fill="currentColor" x="-11" dy="0.32em">80</text>
                </g>
                <g class="tick" opacity="1" transform="translate(0,0.5)">
                  <line stroke="currentColor" x2="-6"></line><text fill="currentColor" x="-11" dy="0.32em">100</text>
                </g>
              </g><text class="intensity-axis-label" transform="translate(-30, 120) rotate(-90)">Relative intensity</text><text style="text-anchor: middle;" transform="translate(60, 270) ">m/z</text>
              <g class="sample-graph soloSample">
                <g>
                  <circle cx="9.999999999999526" cy="0" r="4"></circle>
                  <circle cx="110.00000000000047" cy="96" r="4"></circle>
                </g>
                <g>
                  <line x1="9.999999999999526" x2="9.999999999999526" y1="0" y2="240"></line>
                  <line x1="110.00000000000047" x2="110.00000000000047" y1="96" y2="240"></line>
                </g>
                <g>
                  <rect width="0.0599999999991514" height="240" x="9.969999999999951" y="0"></rect>
                  <rect width="0.06060000000047694" height="144" x="109.96970000000023" y="96"></rect>
                </g>
              </g>
              <g class="theor-graph soloTheor">
                <g>
                  <path d="M0,240L9.999999999999526,237.6L19.99999999999905,240L100.00000000000095,240L110.00000000000047,238.8L120,240"></path>
                </g>
              </g>
              <g fill="none" pointer-events="all">
                <rect class="overlay" pointer-events="all" cursor="crosshair" x="0" y="0" width="120" height="240"></rect>
                <rect class="selection" cursor="move" fill="#777" fill-opacity="0.3" stroke="#fff" shape-rendering="crispEdges" style="display: none;"></rect>
                <rect class="handle handle--w" cursor="ew-resize" style="display: none;"></rect>
                <rect class="handle handle--e" cursor="ew-resize" style="display: none;"></rect>
              </g>
            </g>
          </svg></div>
        <div class="plot-legend">
          <div class="legend-item"><svg width="64" height="32">
              <g>
                <g class="theor-graph soloTheor">
                  <g>
                    <path d="M0,31L2,31L4,31L6,31L8,31L10,30.7L12,30.4L14,29.8L16,28.3L18,26.2L20,22.9L22,18.4L24,13.3L26,7.900000000000002L28,3.6999999999999993L30,1L32,1L34,3.400000000000002L36,7.900000000000002L38,13L40,18.4L42,22.6L44,26.2L46,28.3L48,29.5L50,30.4L52,30.7L54,31L56,31L58,31L60,31L62,31"></path>
                  </g>
                </g>
              </g>
            </svg> <span class="legend-item-name">Theoretical</span></div>
          <div class="legend-item"><svg width="64" height="32">
              <g>
                <g class="sample-graph soloSample">
                  <g>
                    <circle cx="32" cy="4" r="4"></circle>
                  </g>
                  <g>
                    <line x1="32" x2="32" y1="4" y2="32"></line>
                  </g>
                  <g>
                    <rect width="48" height="28" x="8" y="4"></rect>
                  </g>
                </g>
              </g>
            </svg> <span class="legend-item-name">Sample</span></div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Diagnostics should match snapshot (with isobar) 1`] = `
<div mock-v-loading="0">
  <transition-stub name="el-alert-fade" class="isobar-alert">
    <div role="alert" class="el-alert el-alert--warning is-light"><i class="el-alert__icon el-icon-warning is-big"></i>
      <div class="el-alert__content">
        <!---->
        <p class="el-alert__description">
          Another ion was annotated that is isobaric to the selected annotation.
          <!---->
          Select an isobaric annotation below to compare against.
        </p>
        <!----><i class="el-alert__closebtn el-icon-close" style="display: none;"></i>
      </div>
    </div>
  </transition-stub>
  <div class="compare-container">
    Compare selected annotation to:
    <mock-el-select placeholder="None" clearable="" class="compare-select" value="C8H9S2">
      <mock-el-option value="C8H9S2" label="[M+0.0000]: [C8H10S2 - H]⁻">[M+0.0000]: [C<sub>8</sub>H<sub>10</sub>S<sub>2</sub> - H]⁻</mock-el-option>
    </mock-el-select>
  </div>
  <div class="ref-annotation-header">
    <mock-el-popover trigger="hover" placement="top" open-delay="100">
      <div slot-key="default">
        <div class="leading-5 py-2 px-3 font-normal content text-sm text-left">
          <!---->
          <p class="title">
            Candidate molecule
          </p>
          <ul class="p-0 list-none leading-7">
            <li>
              C.I. Food Red 6
            </li>
            <!---->
          </ul>
          <!---->
          <!---->
        </div>
      </div>
      <div slot-key="reference"><span class="annotation-ion">[C<sub>19</sub>H<sub>18</sub>N<sub>2</sub>O<sub>7</sub>S<sub>2</sub> - H]⁺</span></div>
    </mock-el-popover> <span style="padding-left: 5px;">(Selected annotation)</span>
  </div>
  <div class="ref-annotation-container">
    <div class="el-row" id="scores-table">
      <div class="msm-score-calc">MSM score = <span>0.657</span> = 0.006 (ρ<sub>spatial</sub>) × 0.791 (ρ<sub>spectral</sub>) × 0.998 (ρ<sub>chaos</sub>)</div>
      <div>
        <mock-el-popover trigger="hover" open-delay="100">
          <div slot-key="default">Image analysis gave an off-sample probability of less than 10%.</div>
          <div slot-key="reference"><span class="on-sample-tag">On-sample</span></div>
        </mock-el-popover>
      </div>
    </div>
    <div class="isotope-images-container tour-isotope-images el-row">
      <div class="el-col el-col-24 el-col-xs-24 el-col-sm-12 el-col-md-12 el-col-lg-6">
        <div class="small-peak-image">
          100.0000
          <mock-el-popover trigger="hover" placement="top">
            <div slot-key="default">
              This peak's detection window overlaps with the other annotation's 1st peak.
            </div>
            <div slot-key="reference"><i class="el-icon-warning-outline"></i></div>
          </mock-el-popover> <br>
          <imageloader-stub src="/fs/iso_images/img1" colormap="Viridis" image-fit-params="[object Object]" min-intensity="0" max-intensity="100" show-pixel-intensity="" annotimageopacity="1" opacitymode="constant" imageposition="[object Object]" pixelaspectratio="1" style="overflow: hidden;"></imageloader-stub>
        </div>
      </div>
      <div class="el-col el-col-24 el-col-xs-24 el-col-sm-12 el-col-md-12 el-col-lg-6">
        <div class="small-peak-image">
          101.0000
          <!----> <br>
          <imageloader-stub src="/fs/iso_images/img2" colormap="Viridis" image-fit-params="[object Object]" min-intensity="100" max-intensity="200" show-pixel-intensity="" annotimageopacity="1" opacitymode="constant" imageposition="[object Object]" pixelaspectratio="1" style="overflow: hidden;"></imageloader-stub>
        </div>
      </div>
    </div>
  </div>
  <div class="comp-annotation-header">
    <!----> <span><mock-el-popover trigger="hover" placement="top" open-delay="100"><div slot-key="default"><div class="leading-5 py-2 px-3 font-normal content text-sm text-left"><!----> <p class="title">
      Candidate molecule
    </p> <ul class="p-0 list-none leading-7"><li>
        C.I. Food Red 6
      </li> <!----></ul> <!----> <!----></div> </div><div slot-key="reference"><span class="annotation-ion">[C<sub>8</sub>H<sub>10</sub>S<sub>2</sub> - H]⁻</span></div>
  </mock-el-popover>
  <!----></span>
</div>
<div class="comp-annotation-container">
  <div class="el-row" id="scores-table">
    <div class="msm-score-calc" mock-v-loading="1">MSM score = <span>0.657</span> = 0.006 (ρ<sub>spatial</sub>) × 0.791 (ρ<sub>spectral</sub>) × 0.998 (ρ<sub>chaos</sub>)</div>
    <div>
      <mock-el-popover trigger="hover" open-delay="100">
        <div slot-key="default">Image analysis gave an off-sample probability of 88%.</div>
        <div slot-key="reference"><span class="off-sample-tag">Off-sample</span></div>
      </mock-el-popover>
    </div>
  </div>
  <div class="isotope-images-container tour-isotope-images el-row">
    <div class="el-col el-col-24 el-col-xs-24 el-col-sm-12 el-col-md-12 el-col-lg-6">
      <div class="small-peak-image">
        100.0000
        <mock-el-popover trigger="hover" placement="top">
          <div slot-key="default">
            This peak's detection window overlaps with the other annotation's 1st peak.
          </div>
          <div slot-key="reference"><i class="el-icon-warning-outline"></i></div>
        </mock-el-popover> <br>
        <imageloader-stub src="/fs/iso_images/img1" colormap="Viridis" image-fit-params="[object Object]" min-intensity="0" max-intensity="100" show-pixel-intensity="" annotimageopacity="1" opacitymode="constant" imageposition="[object Object]" pixelaspectratio="1" style="overflow: hidden;"></imageloader-stub>
      </div>
    </div>
    <div class="el-col el-col-24 el-col-xs-24 el-col-sm-12 el-col-md-12 el-col-lg-6">
      <div class="small-peak-image">
        101.0000
        <!----> <br>
        <imageloader-stub src="/fs/iso_images/img2" colormap="Viridis" image-fit-params="[object Object]" min-intensity="100" max-intensity="200" show-pixel-intensity="" annotimageopacity="1" opacitymode="constant" imageposition="[object Object]" pixelaspectratio="1" style="overflow: hidden;"></imageloader-stub>
      </div>
    </div>
  </div>
</div>
<div>
  <div class="el-row" id="isotope-plot-container">
    <div class="peak-chart">
      <div style="height: 300px;"><svg width="200" height="300">
          <defs>
            <pattern id="pattern-stripe" width="6" height="6" patternUnits="userSpaceOnUse" patternTransform="rotate(45)">
              <rect width="3" height="6" transform="translate(0,0)" fill="white"></rect>
            </pattern>
            <mask id="mask-stripe">
              <rect x="0" y="0" width="100%" height="100%" fill="url(#pattern-stripe)"></rect>
            </mask>
          </defs>
          <g transform="translate(40, 10)">
            <g transform="translate(0, 240)" fill="none" font-size="10" font-family="sans-serif" text-anchor="middle">
              <path class="domain" stroke="currentColor" d="M0.5,6V0.5H120.5V6"></path>
              <g class="tick" opacity="1" transform="translate(10.499999999999526,0)">
                <line stroke="currentColor" y2="6"></line><text fill="currentColor" y="9" dy="0.71em">100.0</text>
              </g>
              <g class="tick" opacity="1" transform="translate(30.5,0)">
                <line stroke="currentColor" y2="6"></line><text fill="currentColor" y="9" dy="0.71em">100.2</text>
              </g>
              <g class="tick" opacity="1" transform="translate(50.500000000000476,0)">
                <line stroke="currentColor" y2="6"></line><text fill="currentColor" y="9" dy="0.71em">100.4</text>
              </g>
              <g class="tick" opacity="1" transform="translate(70.49999999999953,0)">
                <line stroke="currentColor" y2="6"></line><text fill="currentColor" y="9" dy="0.71em">100.6</text>
              </g>
              <g class="tick" opacity="1" transform="translate(90.5,0)">
                <line stroke="currentColor" y2="6"></line><text fill="currentColor" y="9" dy="0.71em">100.8</text>
              </g>
              <g class="tick" opacity="1" transform="translate(110.50000000000047,0)">
                <line stroke="currentColor" y2="6"></line><text fill="currentColor" y="9" dy="0.71em">101.0</text>
              </g>
            </g>
            <g fill="none" font-size="10" font-family="sans-serif" text-anchor="end">
              <path class="domain" stroke="currentColor" d="M-6,240.5H0.5V0.5H-6"></path>
              <g class="tick" opacity="1" transform="translate(0,240.5)">
                <line stroke="currentColor" x2="-6"></line><text fill="currentColor" x="-11" dy="0.32em">0</text>
              </g>
              <g class="tick" opacity="1" transform="translate(0,192.5)">
                <line stroke="currentColor" x2="-6"></line><text fill="currentColor" x="-11" dy="0.32em">20</text>
              </g>
              <g class="tick" opacity="1" transform="translate(0,144.5)">
                <line stroke="currentColor" x2="-6"></line><text fill="currentColor" x="-11" dy="0.32em">40</text>
              </g>
              <g class="tick" opacity="1" transform="translate(0,96.5)">
                <line stroke="currentColor" x2="-6"></line><text fill="currentColor" x="-11" dy="0.32em">60</text>
              </g>
              <g class="tick" opacity="1" transform="translate(0,48.5)">
                <line stroke="currentColor" x2="-6"></line><text fill="currentColor" x="-11" dy="0.32em">80</text>
              </g>
              <g class="tick" opacity="1" transform="translate(0,0.5)">
                <line stroke="currentColor" x2="-6"></line><text fill="currentColor" x="-11" dy="0.32em">100</text>
              </g>
            </g><text class="intensity-axis-label" transform="translate(-30, 120) rotate(-90)">Relative intensity</text><text style="text-anchor: middle;" transform="translate(60, 270) ">m/z</text>
            <g class="sample-graph soloSample">
              <g>
                <circle cx="9.999999999999526" cy="0" r="4"></circle>
                <circle cx="110.00000000000047" cy="96" r="4"></circle>
              </g>
              <g>
                <line x1="9.999999999999526" x2="9.999999999999526" y1="0" y2="240"></line>
                <line x1="110.00000000000047" x2="110.00000000000047" y1="96" y2="240"></line>
              </g>
              <g>
                <rect width="0.0599999999991514" height="240" x="9.969999999999951" y="0"></rect>
                <rect width="0.06060000000047694" height="144" x="109.96970000000023" y="96"></rect>
              </g>
            </g>
            <g class="theor-graph soloTheor">
              <g>
                <path d="M0,240L9.999999999999526,237.6L19.99999999999905,240L100.00000000000095,240L110.00000000000047,238.8L120,240"></path>
              </g>
            </g>
            <g fill="none" pointer-events="all">
              <rect class="overlay" pointer-events="all" cursor="crosshair" x="0" y="0" width="120" height="240"></rect>
              <rect class="selection" cursor="move" fill="#777" fill-opacity="0.3" stroke="#fff" shape-rendering="crispEdges" style="display: none;"></rect>
              <rect class="handle handle--w" cursor="ew-resize" style="display: none;"></rect>
              <rect class="handle handle--e" cursor="ew-resize" style="display: none;"></rect>
            </g>
          </g>
        </svg></div>
      <div class="plot-legend">
        <div class="legend-item"><svg width="64" height="32">
            <g>
              <g class="theor-graph refTheor">
                <g>
                  <path d="M0,31L2,31L4,31L6,31L8,31L10,30.7L12,30.4L14,29.8L16,28.3L18,26.2L20,22.9L22,18.4L24,13.3L26,7.900000000000002L28,3.6999999999999993L30,1L32,1L34,3.400000000000002L36,7.900000000000002L38,13L40,18.4L42,22.6L44,26.2L46,28.3L48,29.5L50,30.4L52,30.7L54,31L56,31L58,31L60,31L62,31"></path>
                </g>
              </g>
            </g>
          </svg> <span class="legend-item-name">[C<sub>19</sub>H<sub>18</sub>N<sub>2</sub>O<sub>7</sub>S<sub>2</sub> - H]⁺<br> Theoretical</span></div>
        <div class="legend-item"><svg width="64" height="32">
            <g>
              <g class="sample-graph refSample">
                <g>
                  <circle cx="32" cy="4" r="4"></circle>
                </g>
                <g>
                  <line x1="32" x2="32" y1="4" y2="32"></line>
                </g>
                <g>
                  <rect width="48" height="28" x="8" y="4"></rect>
                </g>
              </g>
            </g>
          </svg> <span class="legend-item-name">[C<sub>19</sub>H<sub>18</sub>N<sub>2</sub>O<sub>7</sub>S<sub>2</sub> - H]⁺<br> Sample</span></div>
        <div class="legend-item"><svg width="64" height="32">
            <g>
              <g class="theor-graph compTheor">
                <g>
                  <path d="M0,31L2,31L4,31L6,31L8,31L10,30.7L12,30.4L14,29.8L16,28.3L18,26.2L20,22.9L22,18.4L24,13.3L26,7.900000000000002L28,3.6999999999999993L30,1L32,1L34,3.400000000000002L36,7.900000000000002L38,13L40,18.4L42,22.6L44,26.2L46,28.3L48,29.5L50,30.4L52,30.7L54,31L56,31L58,31L60,31L62,31"></path>
                </g>
              </g>
            </g>
          </svg> <span class="legend-item-name">[C<sub>8</sub>H<sub>10</sub>S<sub>2</sub> - H]⁻<br> Theoretical</span></div>
        <div class="legend-item"><svg width="64" height="32">
            <g>
              <g class="sample-graph compSample">
                <g>
                  <circle cx="32" cy="4" r="4"></circle>
                </g>
                <g>
                  <line x1="32" x2="32" y1="4" y2="32"></line>
                </g>
                <g>
                  <rect width="48" height="28" x="8" y="4"></rect>
                </g>
              </g>
            </g>
          </svg> <span class="legend-item-name">[C<sub>8</sub>H<sub>10</sub>S<sub>2</sub> - H]⁻<br> Sample</span></div>
      </div>
    </div>
  </div>
</div>
</div>
`;
